### PR TITLE
Milestone-1 TCP-Transport. 

### PR DESCRIPTION
**Pull request by ayuryshev**
Excluding the last task defined  in https://github.com/skycoin/skywire/issues/501 - everything is completed.
It was already done by most part a long time before.
The most hard part is in the last task.

### CODE REVIEW

File: pkg/snet/network.go [hunk 1]
```diff
@@ -39,16 +40,22 @@ type Config struct {
 
 	DmsgDiscAddr string
 	DmsgMinSrvs  int
+
+	LocalTCPAddress string
```
Comment (evanlinjin) [inline pkg/snet/network.go]: Rename to `TCPLocalAddr`

File: pkg/snet/network.go [hunk 2]
```diff
@@ -39,16 +40,22 @@ type Config struct {
...
+	PubKeyFile      string
```
Comment (evanlinjin) [inline pkg/snet/network.go]: Rename to `TCPPubKeyFile`

File: pkg/snet/network.go [hunk 3]
```diff
@@ -26,6 +26,7 @@ const (
 // Network types.
 const (
 	DmsgType = "dmsg"
+	TCPType  = "tcptr"
```
Comment (evanlinjin) [inline pkg/snet/network.go]: Should be `tcp`

File: pkg/snet/network.go [hunk 4]
```diff
@@ -118,9 +133,33 @@ func (n *Network) Listen(network string, port uint16) (*Listener, error) {
...
+				network:  TCPType,
+			}, nil
+		}
 	default:
 		return nil, ErrUnknownNetwork
 	}
+	return nil, nil
+}
+
+func (n *Network) TCP() *TCPFactory {
```
Comment (evanlinjin) [inline pkg/snet/network.go]: Put this on line `107` to keep things orderly.

File: pkg/snet/tcp_transport.go [hunk 5]
```diff
@@ -1,9 +1,12 @@
-package transport
+// +build !no_ci
```
Comment (evanlinjin) [inline pkg/snet/tcp_transport.go]: Rename file to `tcp.go`. `Transport` is no longer an interface, but something that wraps `net.Conn`.

File: pkg/snet/tcp_transport.go [hunk 6]
```diff
@@ -101,35 +119,52 @@ func (tr *TCPTransport) Type() string {
...
+func (t *memPKTable) RemoteAddr(remotePK cipher.PubKey) string {
 	return t.entries[remotePK]
 }
 
-func (t *inMemoryPKTable) RemotePK(remoteIP net.IP) cipher.PubKey {
-	for pk, addr := range t.entries {
-		if addr.IP.String() == remoteIP.String() {
-			return pk
-		}
+func (t *memPKTable) RemotePK(address string) cipher.PubKey {
```
Comment (evanlinjin) [inline pkg/snet/tcp_transport.go]: Have this return `(cipher.PubKey, bool)`

File: pkg/snet/tcp_transport.go [hunk 7]
```diff
@@ -74,7 +92,7 @@ func (f *TCPFactory) Local() cipher.PubKey {
 
 // Type returns the Transport type.
 func (f *TCPFactory) Type() string {
-	return "tcp"
+	return "tcp-transport"
```
Comment (evanlinjin) [inline pkg/snet/tcp_transport.go]: `"tcp"` is fine. Better to use the defined constant though.

File: pkg/snet/tcp_transport.go (hunk 5 above)
Comment (evanlinjin) [reply pkg/snet/tcp_transport.go]: I think it is best to put this in a separate package `/pkg/snet/stcp`.

File: pkg/snet/tcp_transport.go [hunk 8]
```diff
@@ -101,35 +119,52 @@ func (tr *TCPTransport) Type() string {
 
 // PubKeyTable provides translation between remote PubKey and TCPAddr.
 type PubKeyTable interface {
-	RemoteAddr(remotePK cipher.PubKey) *net.TCPAddr
-	RemotePK(remoteIP net.IP) cipher.PubKey
+	RemoteAddr(remotePK cipher.PubKey) string
```
Comment (evanlinjin) [inline pkg/snet/tcp_transport.go]: Why have we decided to use `string` instead of `net.TCPAddr`?

File: pkg/snet/tcp_transport.go [hunk 9]
```diff
@@ -101,35 +119,52 @@ func (tr *TCPTransport) Type() string {
...
+}
+
+type memPKTable struct {
+	entries map[cipher.PubKey]string
+	reverse map[string]cipher.PubKey
 }
 
-type inMemoryPKTable struct {
-	entries map[cipher.PubKey]*net.TCPAddr
+func memoryPubKeyTable(entries map[cipher.PubKey]string) *memPKTable {
```
Comment (evanlinjin) [inline pkg/snet/tcp_transport.go]: Rename to `makeMemPKTable` to be consistent.

File: pkg/snet/tcp_transport.go [hunk 10]
```diff
@@ -101,35 +119,52 @@ func (tr *TCPTransport) Type() string {
...
+	reverse := make(map[string]cipher.PubKey)
+	for k, v := range entries {
+		addr, err := net.ResolveTCPAddr("tcp", v)
+		if err != nil {
+			panic("error in resolving address")
+		}
+		reverse[addr.IP.String()] = k
```
Comment (evanlinjin) [inline pkg/snet/tcp_transport.go]: Why are we using purely IP? Can we just use tcp address with `host:port`?

File: pkg/snet/tcp_transport.go [hunk 11]
```diff
@@ -101,35 +119,52 @@ func (tr *TCPTransport) Type() string {
...
+	}
+	return &memPKTable{entries, reverse}
 }
 
-// InMemoryPubKeyTable returns in memory implementation of the PubKeyTable.
-func InMemoryPubKeyTable(entries map[cipher.PubKey]*net.TCPAddr) PubKeyTable {
-	return &inMemoryPKTable{entries}
+// MemoryPubKeyTable returns in memory implementation of the PubKeyTable.
+func MemoryPubKeyTable(entries map[cipher.PubKey]string) PubKeyTable {
```
Comment (evanlinjin) [inline pkg/snet/tcp_transport.go]: `MakeMemoryPubKeyTable`

**Review by evanlinjin — CHANGES_REQUESTED** [review]

File: pkg/snet/tcp_transport.go (hunk 10 above)
Comment (ayuryshev) [reply pkg/snet/tcp_transport.go]: It was my first thought too.
But then I remembered/encountered that for dialling side an ephemereal port is assigned. 
In the case when N nodes are dialling to single node that accepts connections - all diallers will have different ephemereal port.
So PK<==>IP  is simplest (and maybe the only practical) decision.

File: pkg/snet/tcp_transport.go (hunk 5 above)
Comment (ayuryshev) [reply pkg/snet/tcp_transport.go]: Renamed.
About separate package: `net` from stdlib uses only one nesting level and covers all types of TCP/IP. I propose to stick with this example.

File: pkg/snet/tcp_transport.go (hunk 8 above)
Comment (ayuryshev) [reply pkg/snet/tcp_transport.go]: No reason :)
I just forgot to change signature after merge from mainnet
Although reasons for using address in string forms exists. 
Take a look how  became:
```golang
func (t *memPKTable) RemoteAddr(remotePK cipher.PubKey) *net.TCPAddr {
	rAddr, _ := net.ResolveTCPAddr("tcp", t.entries[remotePK]) // nolint: errcheck
	return rAddr
}
```

File: pkg/snet/tcp_transport.go (hunk 6 above)
Comment (ayuryshev) [reply pkg/snet/tcp_transport.go]: OK

**Comment by ayuryshev** [comment]
Fixes requested by Evan - done.
Linter errors and failed tests that were before this PR remains the same.

**Comment by ayuryshev** [comment]
Remerged with mainnet-milestone-1. Fixed tests for tcp-transport

**Pull request closed without merging**